nt-be/src/handlers/balance_changes/account_monitor.rs

```diff
@@ -0,0 +1,367 @@
+use near_api::NetworkConfig;
+use sqlx::PgPool;
+use std::collections::HashSet;
+
+use super::balance::ft::get_balance_at_block as get_ft_balance;
+use super::gap_filler::{fill_gaps, insert_snapshot_record};
+use super::token_discovery::snapshot_intents_tokens;
+
+/// Run one cycle of monitoring for all enabled accounts
+///
+/// This function:
+/// 1. Queries all enabled accounts from monitored_accounts table
+/// 2. For each account:
+///    - Gets all known tokens for that account from balance_changes
+///    - Runs gap filling for each token up to the specified block
+///    - Updates last_synced_at timestamp after processing
+/// 3. Handles errors gracefully, continuing with next account if one fails
+pub async fn run_monitor_cycle(
+    pool: &PgPool,
+    network: &NetworkConfig,
+    up_to_block: i64,
+) -> Result<(), Box<dyn std::error::Error>> {
+    // Get all enabled monitored accounts
+    let accounts = sqlx::query!(
+        r#"
+        SELECT account_id, last_synced_at
+        FROM monitored_accounts
+        WHERE enabled = true
+        ORDER BY 
+            CASE WHEN last_synced_at IS NULL THEN 0 ELSE 1 END,
+            last_synced_at ASC NULLS FIRST
+        "#
+    )
+    .fetch_all(pool)
+    .await?;
+
+    if accounts.is_empty() {
+        println!("No enabled accounts to monitor");
+        return Ok(());
+    }
+
+    println!("Monitoring {} enabled accounts", accounts.len());
+
+    for account in accounts {
+        let account_id = &account.account_id;
+
+        // Get all unique tokens for this account (excluding nulls which shouldn't happen but be safe)
+        let mut tokens: Vec<String> = sqlx::query_scalar(
```

Comment:
This is n+1 query, and it doesn't scale well. Instead, you can use join in the  above query and process a loop as Account + Token